src/topology/builder.rs

```diff
@@ -152,33 +166,64 @@ pub async fn build_pieces(
         let typetag = sink.inner.sink_type();
         let input_type = sink.inner.input_type();
 
-        let buffer = sink.buffer.build(&config.global.data_dir, &name);
-        let (tx, rx, acker) = match buffer {
-            Err(error) => {
-                errors.push(format!("Sink \"{}\": {}", name, error));
-                continue;
+        let (tx, rx, acker) = if let Some(buffer) = buffers.remove(name) {
+            buffer
+        } else {
+            let buffer = sink.buffer.build(&config.global.data_dir, &name);
+            match buffer {
+                Err(error) => {
+                    errors.push(format!("Sink \"{}\": {}", name, error));
+                    continue;
+                }
+                Ok((tx, rx, acker)) => (tx, Arc::new(Mutex::new(Some(rx))), acker),
             }
-            Ok(buffer) => buffer,
         };
 
-        let cx = SinkContext { acker };
+        let cx = SinkContext {
+            acker: acker.clone(),
+        };
 
         let (sink, healthcheck) = match sink.inner.build(cx).await {
             Err(error) => {
                 errors.push(format!("Sink \"{}\": {}", name, error));
                 continue;
             }
-            Ok((sink, healthcheck)) => (sink, healthcheck),
+            Ok(builded) => builded,
         };
 
-        let sink = sink
-            .run(
-                filter_event_type(rx, input_type)
+        let (trigger, tripwire) = Tripwire::new();
+
+        let sink = async move {
+            // Why is this Arc<Mutex<Option<_>>> needed you ask.
+            // In case when this function build_pieces errors
+            // this future won't be run so this rx won't be taken
+            // which will enable us to reuse rx to rebuild
+            // old configuration by passing this Arc<Mutex<Option<_>>>
+            // yet again.
```

Comment:
How will we pass it again? I see that it won't be taken unless the sink is actually run, but it does seem that every time we call `build_pieces` it is either `remove`'d from `buffers` of newly built. Either way, it doesn't seem like we would still have it if we fail and call the function again.

Reply:
That's all true, and the missing piece is that we are cloning `buffers` when calling `build_pieces`. 
https://github.com/timberio/vector/blob/437cb648fe45adb5cc9ab20739b5c8b766b0d0e0/src/topology/mod.rs#L248